src/Microsoft.AspNetCore.Razor.VSCode/src/RazorProjectManager.ts

```diff
@@ -0,0 +1,172 @@
+import { getUriPath } from './UriPaths';
+
+const configurationFileGlobbingPath = `**/project.razor.json`;
+const csprojGlobbingPath = `**/*.csproj`;
+
+export class RazorProjectManager {
+    private readonly razorProjects: { [csprojPath: string]: RazorProject } = {};
+    private onChangeEmitter = new vscode.EventEmitter<IRazorProjectChangeEvent>();
+
+    constructor(private readonly logger: RazorLogger) {
+    }
+
+    public get onChange() { return this.onChangeEmitter.event; }
+
+    public async initialize() {
+        // Track current projects
+        const projectUris = await vscode.workspace.findFiles(csprojGlobbingPath);
+        for (const uri of projectUris) {
+            this.addProject(uri);
+        }
+
+        const projectConfigurationUris = await vscode.workspace.findFiles(configurationFileGlobbingPath);
+        for (const configurationUri of projectConfigurationUris) {
+            this.updateProjectConfiguration(configurationUri);
+        }
+    }
+
+    public register() {
+        // Track future projects
+        const projectWatcher = vscode.workspace.createFileSystemWatcher(csprojGlobbingPath);
+        const didCreateRegistration = projectWatcher.onDidCreate(
+            async (uri: vscode.Uri) => this.addProject(uri));
+        const didDeleteRegistration = projectWatcher.onDidDelete(
+            async (uri: vscode.Uri) => this.removeProject(uri));
+
+        // Track future projects data files
+        const configurationWatcher = vscode.workspace.createFileSystemWatcher(configurationFileGlobbingPath);
+        const didCreateConfigRegistration = configurationWatcher.onDidCreate(
+            async (uri: vscode.Uri) => this.updateProjectConfiguration(uri));
+        const didDeleteConfigRegistration = configurationWatcher.onDidDelete(
+            async (uri: vscode.Uri) => this.removeProjectConfiguration(uri));
+        const didChangeConfigRegistration = configurationWatcher.onDidChange(
+            async (uri: vscode.Uri) => this.updateProjectConfiguration(uri));
+
+        return vscode.Disposable.from(
+            configurationWatcher,
+            didCreateRegistration,
+            didDeleteRegistration,
+            didCreateConfigRegistration,
+            didDeleteConfigRegistration,
+            didChangeConfigRegistration);
+    }
+
+    private updateProjectConfiguration(configurationUri: vscode.Uri) {
+        const configuration = this.getProjectConfiguration(configurationUri);
+        if (!configuration) {
+            return;
+        }
+
+        if (!this.razorProjects[configuration.projectPath]) {
+            this.logger.logVerbose(
+                `Invalid project config. Could not find a corresponding project for ${configuration.projectPath}`);
+            return;
+        }
+
+        const projectContainer = this._getProject(configuration.projectUri);
+        const newConfiguration = { ...projectContainer.configuration };
```

Comment:
Could you clarify why it's necessary to clone the earlier `configuration` dict before editing it? It looks like we're overwriting the old `this.razorProjects` entry entirely anyway, so I'm not sure why we can't just mutate it in place. Are the `configuration` objects somehow shared by multiple projects sometimes?

Reply:
We could mutate it in place; however, i'm trying to correct one of the mistakes I made with the `RazorDocumentManager`. Having immutable projects enforces consumers to rely on the change event from this class rather than holding onto the projects. That way we have less coupling and have more flexibility on what we can and cannot change in regards to the Razor projects.

This shouldn't happen frequently so it's safer to make things immutable.